drivers/service_webhook.go

```diff
@@ -0,0 +1,120 @@
+package drivers
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	v1client "github.com/rancher/go-rancher/client"
+	"github.com/rancher/go-rancher/v2"
+
+	"github.com/rancher/webhook-service/config"
+	"github.com/rancher/webhook-service/model"
+)
+
+type ServiceWebhookDriver struct {
+}
+
+func (s *ServiceWebhookDriver) ValidatePayload(conf interface{}, apiClient *client.RancherClient) (int, error) {
+	if _, ok := conf.(model.ServiceWebhook); !ok {
+		return http.StatusInternalServerError, fmt.Errorf("Can't process config")
+	}
+	return http.StatusOK, nil
+}
+
+func (s *ServiceWebhookDriver) Execute(conf interface{}, apiClient *client.RancherClient, request *http.Request) (int, error) {
+	requestPayloadByte, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		return 500, err
+	}
+	rancherConfig := config.GetConfig()
+	webhookConfig := &model.ServiceWebhook{}
+	if err = mapstructure.Decode(conf, webhookConfig); err != nil {
+		return http.StatusInternalServerError, errors.Wrap(err, "Couldn't unmarshal config")
+	}
+
+	arry := strings.Split(request.RequestURI, "?")
+	projectID, errCode, err := getProjectID(request)
+	if err != nil {
+		return errCode, err
+	}
+	CattleAddr := rancherConfig.CattleURL[:len(rancherConfig.CattleURL)-3]
+	log.Debugf("Excute rancherConfig.CattleURL %v", CattleAddr)
+	postURL := fmt.Sprintf("%s/r/projects/%s/%s:%s%s", CattleAddr, projectID, webhookConfig.ServiceName, webhookConfig.Port, webhookConfig.Path)
+
+	// append the query parameters to the postURL
+	if arry[1] != "" {
+		postURL += "?" + arry[1]
+	}
+	log.Debugf("Excute postURL %v", postURL)
+	log.Debugf("Excute requestPayloadByte %v", requestPayloadByte)
+	hopRequest, err := http.NewRequest("POST", postURL, bytes.NewBuffer(requestPayloadByte))
+	if err != nil {
+		log.Errorf("Fail: %v", err)
+		return http.StatusInternalServerError, err
+	}
+
+	client := &http.Client{}
+	hopRequest.Header = request.Header
+	hopRequest.SetBasicAuth(rancherConfig.CattleAccessKey, rancherConfig.CattleSecretKey)
+	resp, err := client.Do(hopRequest)
+	if err != nil {
+		log.Errorf("Error sending request to service:%v", err)
```

Comment:
We like to avoid logging an error and then also returning that error. You should only log the error if you are handling it and not bubbling it up. Otherwise, it it is the responsibility of the caller to log the error.

Reply:
@cjellick Ok, I will only return the error.